### ISSUE TITLE
[Question] SKRL - How to add new parameters in tensorboard logging? (managed RL)

### ISSUE THREAD

**Issue by AndreaRossetto** [isaac-sim/IsaacLab#3492]
### Question

Hi,
I need to add new log values during training, I have all the rewards but I need also to log other things, for example the joints torque but I'm not able to find how I can do it.

<img width="3386" height="1905" alt="Image" src="https://github.com/user-attachments/assets/637b5c9e-4847-47e6-94ff-6045499b0426" />

Where I can find an example code to do that?

**Comment by RandomOakForest** [comment]
Thank you for posting this. Have you tried `track_data`? SKRL allows tracking of any custom scalar metric by simply passing the name and value to `track_data()`. For example,
```python
agent.track_data("Torques / mean_joint_torque", mean_joint_torque)
```
or
```python
agent.writer.add_scalar("Torques / joint_torque", joint_torque_value, global_step=step)
```

**Comment by AndreaRossetto** [comment]
I've found it but I don't know where to put it. I've to create an CurrTerm() with that funcion or I have to modify something else? where I find the "agent" in a Managed env?

**Comment by Toni-SM** [comment]
Hi @AndreaRossetto 

I don't know manager-based workflow in detail, but similar to the direct workflow, the is the `self.extras["log"]` dict is used to log additional/custom information to TensorBoard. 

https://github.com/isaac-sim/IsaacLab/blob/3a0db9d761982dc65417e6c6d0714cec61ceadb3/source/isaaclab/isaaclab/envs/manager_based_rl_env.py#L367

Now, we need to ask to a manager-based expert about how to access it from the task config.
@Mayankm96 @kellyguo11 @ooctipus for viz

**Comment by AndreaRossetto** [comment]
@Mayankm96  @kellyguo11 @ooctipus  No one can help me ?

**Comment by RandomOakForest** [comment]
Our apologies for the delay. The team has been caught up with conferences and whatnot. I'll move this post to our _Discussions_ for follow up. Here are a few possible next steps.

When using a **Manager-Based RL environment** (via `ManagerBasedRLEnv` in Isaac Lab) with **SKRL**, you normally don’t create a `CurrTerm()` or similar; instead, you hook your custom logging into the `agent` that’s running. However, since in a managed training setup (like those launched from `scripts/reinforcement_learning/skrl/train.py`), the `agent` is **not defined inside the environment**, you must access it through the **SKRL trainer** script rather than modifying the environment itself.

Here’s how to integrate your custom logging like `track_data()` in this setup:

### 1. Where the `agent` lives

In Isaac Lab’s SKRL integration, the **agent** is instantiated in the training script, not inside your environment configuration.
If you open the standard SKRL training script (refer to `scripts/reinforcement_learning/skrl/train.py`), you’ll find something like:

```python
from skrl.trainers.torch import SequentialTrainer
from skrl.agents.torch.ppo import PPO

agent = PPO(models=models,
             memory=memory,
             cfg=agent_cfg,
             observation_space=env.observation_space,
             action_space=env.action_space,
             device=env.device)

trainer = SequentialTrainer(env=env, agents=agent, cfg=train_cfg)
trainer.train()
```

This means your **`agent` object** is the `PPO` instance (or another algorithm) passed into the `SequentialTrainer`.


### 2. How to Add Custom Logs

You can inject your custom logging code in one of these ways:

#### **A. Modify the Trainer Loop**

If you have access to the SKRL trainer code (e.g., you cloned Isaac Lab and run your own training scripts), you can subclass the trainer and add custom hooks:

```python
from skrl.trainers.torch import SequentialTrainer

class CustomTrainer(SequentialTrainer):
    def train(self):
        while self.progress < 1.0:
            # perform the usual SKRL training step
            super().train_step()

            # your custom log data
            torques = self.env.scene.data.applied_torque.mean().item()
            self.agents[0].track_data("Torques / mean_joint_torque", torques)
```

Then call:

```python
trainer = CustomTrainer(env=env, agents=[agent], cfg=cfg)
trainer.train()
```



#### **B. Inside a Reward Term (less typical)**

If you only want to log something from environment internals without changing the trainer, you can instrument your reward function or observation update (inside your task’s MDP file):

```python
def my_custom_reward(env, **kwargs):
    torques = env.scene.data.applied_torque.mean().item()
    env.agent.track_data("Torques / mean_joint_torque", torques)
    return 0.0
```

However, this approach only works if your environment has access to the agent (which a ManagerBasedRLEnv typically doesn’t by default). For reliability, prefer modifying or subclassing the trainer.



### 3. Cleaner Option: Use `Callbacks`

SKRL supports experiment callbacks you can define in your agent configuration (`cfg`).
You can define one like this:

```python
def log_joint_torques(agent, timestep, timesteps):
    torque_mean = agent.env.scene.data.applied_torque.mean().item()
    agent.track_data("Torques / mean_joint_torque", torque_mean)

agent_cfg["experiment"]["callbacks"] = {"on_step_end": [log_joint_torques]}
```

This will automatically call your `log_joint_torques` function every SKRL step, giving access to both the environment and the agent contexts.

***

### Summary

| Method | Recommended For | Access to `agent` | Code location |
| :-- | :-- | :-- | :-- |
| `track_data()` in trainer loop | Most flexible | Yes | Modify SKRL trainer or subclass |
| `track_data()` in reward term | Quick experiments | Sometimes | Inside `mdp.py` or `RewTermCfg` |
| Callback config (`agent_cfg["experiment"]["callbacks"]`) | Clean and modular | Yes | In `agent_cfg` before trainer creation |



This setup is compatible with Isaac Lab version ≥2025.09 and SKRL 1.4.3, based on NVIDIA’s official documentation for **manager-based RL environments** and **SKRL integration** within Isaac Lab.[^1][^2]



[^1]: https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/create_manager_rl_env.html

[^2]: https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_existing_scripts.html

[^3]: https://skrl.readthedocs.io/en/latest/api/envs/isaaclab.html

[^4]: https://isaac-sim.github.io/IsaacLab/main/source/setup/quickstart.html

[^5]: https://forums.developer.nvidia.com/t/skrl-a-modular-reinforcement-learning-library-with-isaac-gym-environments-support/200849

[^6]: https://skrl.readthedocs.io/en/latest/intro/getting_started.html

[^7]: https://isaac-sim.github.io/IsaacLab/main/source/tutorials/03_envs/run_rl_training.html

[^8]: https://docs.robotsfan.com/isaaclab_official/v1.1.0/source/tutorials/03_envs/run_rl_training.html

[^9]: https://github.com/isaac-sim/IsaacLab/issues/3017

[^10]: https://skrl.readthedocs.io/en/latest/intro/examples.html

[^11]: https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/training_guide.html

[^12]: https://forums.developer.nvidia.com/t/trouble-with-reinforcement-learning-task-in-isaac-lab/304145

[^13]: https://docs.robotsfan.com/isaaclab_official/main/_modules/isaaclab_rl/skrl.html

[^14]: https://skrl.readthedocs.io/en/latest/intro/data.html

[^15]: https://www.youtube.com/watch?v=BSQEYj3Wm0Q

[^16]: https://forums.developer.nvidia.com/t/ppo-for-rl-games-vs-skrl/262682

[^17]: https://github.com/Toni-SM/skrl/discussions/298

[^18]: https://skrl.readthedocs.io/en/develop/

[^19]: https://forums.developer.nvidia.com/t/issue-deploying-trained-rl-policies-in-isaac-sim-policy-pt-from-isaac-lab/326021
